### PR TITLE
Adds Make generator

### LINKED ISSUE #3773: GNU Make generator
https://github.com/solvingj/conan-make_generator/blob/master/conanfile.py by @solvingj is almost it.

I agree it could be built-in.

Can use conditional:

```
ifneq ($(USE_CONAN),)
INC_PATHS          += $(CONAN_INC_PATHS)
LD_PATHS           += $(CONAN_LIB_PATHS)
LD_LIBS            += $(CONAN_LIBS)
CXXFLAGS           += $(CONAN_CPP_FLAGS) 
CFLAGS             += $(CONAN_CFLAGS) 
DEFINES            += $(CONAN_DEFINES) 
LDFLAGS_SHARED     += $(CONAN_SHAREDLINKFLAGS) 
LDFLAGS_EXE        += $(CONAN_EXELINKFLAGS)
C_SRCS             += $(CONAN_C_SRCS)
CXX_SRCS           += $(CONAN_CXX_SRCS)
endif
```

### PR DESCRIPTION
**Pull request by danimtb**
Changelog: Feature: Added new ``make`` generator.

- [x] Refer to the issue that supports this Pull Request: closes #3773 
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

**Comment by danimtb** [comment]
I will highly appreciate a review of @solvingj as he is the genuine creator of this generator 😄

File: conans/test/generators/make_test.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+        client.run("new myhello/1.0.0 --sources")
+        conanfile_path = os.path.join(client.current_folder, "conanfile.py")
+        replace_in_file(conanfile_path, "{\"shared\": [True, False]}",
+                        "{\"shared\": [True, False], \"fPIC\": [True, False]}")
+        replace_in_file(conanfile_path, "\"shared=False\"", "\"shared=False\", \"fPIC=True\"")
+        client.run("create . danimtb/testing")
+        hellowrapper_include = """
+#pragma once
+
+#ifdef WIN32
```
Comment (memsharded) [inline conans/test/generators/make_test.py]: Unnecessary if only Linux test

File: conans/client/generators/make.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    @property
+    def makefile_newline(self):
+        return "\n"
+
+    @property
+    def makefile_line_continuation(self):
+        return " \\\n"
+
+    @property
+    def assignment_immediate(self):
```
Comment (memsharded) [inline conans/client/generators/make.py]: Unused?

File: conans/client/generators/__init__.py [hunk 3]
```diff
@@ -28,6 +28,7 @@
 from conans.util.env_reader import get_env
 from .b2 import B2Generator
 from .premake import PremakeGenerator
+from .make import MakeGenerator
```
Comment (SSE4) [inline conans/client/generators/__init__.py]: to be honest, I don't think generic name `MakeGenerator` is the best choice
I suppose it's better to make it e.g. `GNUMakeGenerator` to indicate it works for GNU Make.
(you know, there are lots of incompatible make derivatives, like BSD Make, NMake, etc. and so far it's not tested with them all, but only GNU Make, right?)

File: conans/client/generators/make.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        for line_as_list in self.create_deps_content():
+            content.append("".join(line_as_list))
+
+        content.append("#-------------------------------------------------------------------#")
+        content.append(self.makefile_newline)
+        return self.makefile_newline.join(content)
+
+    def create_deps_content(self):
+        deps_content = []
```
Comment (SSE4) [inline conans/client/generators/make.py]: can be abbreviated to:
```
deps_content = self.create_content_from_deps()
deps_content.extend(self.create_combined_content())
```

File: conans/client/generators/make.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
+from conans.model import Generator
+
+
+class MakeGenerator(Generator):
+
+    @property
+    def filename(self):
+        return "conanbuildinfo.mak"
```
Comment (SSE4) [inline conans/client/generators/make.py]: we usually define constants for these files, may be add one?

File: conans/client/generators/__init__.py (hunk 3 above)
Comment (solvingj) [reply conans/client/generators/__init__.py]: On second thought, this raises an interesting point.  Because this generator only assigns values to variables using the most basic of operators, I believe it might be universally compatible.  Internally, we have a `gnu_make_generator` and `opus_make_generator` which are VERY different, and produce rules and conditionals which are incompatible with each other, but both support the basic assignment operators. 

As such, I now actually think `MakeGenerator` might be the most appropriate name.

Comment (danimtb) [reply conans/client/generators/__init__.py]: - The internal name of the generator class could be changed in the future, what it is important here is the name of the user interface. Of this one it would be ``generators = "make"``. Should we provide a different one no avoid confusion? ``gnumake``? ``generic_make``?

- @solvingj Are you saying that this generator not compatible with GNU Make?

Comment (solvingj) [reply conans/client/generators/__init__.py]: sorry to be unclear.  

This does work with GNU Make.  I was saying that in theory, it should also work for any Make variant. 

RE: naming, I was referring to the generators name for usage.

`make_generic` is probably what i would recommend right now, but i have to say i don't like it, or any of the other options.  I could also accept simply `make` since it really is the most generic and universal `make` generator there will ever be.  Also, it's actually NOT specific to `gnu_make`, as it's known to work with multiple makes (tested with Opus Make).  So, `gnu_make` seems wrong. 
 It would be nice to prove it works with `NMake` in some future test, but that's not required.

So, in summary:  `make` or `generic_make` or `make_generic` all seem fine.  Now i'm leaning towards simply `make` again.

I'll leave it up to Conan team to decide :)

Comment (danimtb) [reply conans/client/generators/__init__.py]: We will leave it as is. Thanks a lot for the feedback and clarification 😄

File: conans/client/generators/make.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    def makefile_line_continuation(self):
+        return " \\\n"
+
+    @property
+    def assignment_if_absent(self):
+        return " ?= "
+
+    @property
+    def assignment_append(self):
+        return " += "
```
Comment (jgsogo) [inline conans/client/generators/make.py]: Why properties and not plain attributes?

File: conans/test/generators/make_test.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+                      "$(CONAN_SHAREDLINKFLAGS_MYPKG1) \\\n"
+                      "$(CONAN_SHAREDLINKFLAGS_MYPKG2)", content)
+        self.assertIn("CONAN_SHAREDLINKFLAGS_MYPKG1 +=  \\\n-framework Cocoa", content)
+        self.assertIn("CONAN_SHAREDLINKFLAGS_MYPKG2 +=  \\\n-framework AudioFoundation", content)
+
+        self.assertIn("CONAN_EXELINKFLAGS +=  \\\n"
+                      "$(CONAN_EXELINKFLAGS_MYPKG1) \\\n"
+                      "$(CONAN_EXELINKFLAGS_MYPKG2)", content)
+        self.assertIn("CONAN_EXELINKFLAGS_MYPKG1 +=  \\\n-framework QuartzCore", content)
+        self.assertIn("CONAN_EXELINKFLAGS_MYPKG2 +=  \\\n-framework VideoToolbox", content)
```
Comment (jgsogo) [inline conans/test/generators/make_test.py]: I prefer to see all the file written and compare with it... all these previous _asserts_ are checking for almost it all, aren't they?

Comment (danimtb) [reply conans/test/generators/make_test.py]: ok, I could change it. Took it from the premake test bu I agree it would be easier to read

File: conans/client/generators/make.py (hunk 6 above)
Comment (danimtb) [reply conans/client/generators/make.py]: I left it as it was. I think it is done that way to prevent any change in the content

File: conans/test/generators/make_test.py (hunk 7 above)
Comment (solvingj) [reply conans/test/generators/make_test.py]: i was thinking about this at end of day yesterday.  Please feel free to refactor the operators in any way that makes sense.  Looking at the current state of the generator, it's possible you can just inline the "assignment" operator text wherever it's used and eliminate the two properties altogether.
